From d34be784b2c3c018628c4f7a0911969fb9b6ef89 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Giuseppe=20Cal=C3=A0?= <jiveaxe@gmail.com>
Date: Sun, 5 Apr 2020 10:53:31 +0200
Subject: [PATCH] Use avatars in FacesDir first and if not found search other
 locations

If Avatars are enabled and home is encrypted it takes 60s to popup
the greeter at boot. The same is if copying $USER.face.icon in FacesDir
since avatar is searched first in home.

The patch changes this logic. Now it search first in FacesDir and if
found the greeter is shown, otherwise queries other default locations.
---
 src/greeter/UserModel.cpp | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/src/greeter/UserModel.cpp b/src/greeter/UserModel.cpp
index 7d7f9e2..6bc25e8 100644
--- a/src/greeter/UserModel.cpp
+++ b/src/greeter/UserModel.cpp
@@ -136,14 +136,16 @@ namespace SDDM {
             if (avatarsEnabled) {
                 const QString userFace = QStringLiteral("%1/.face.icon").arg(user->homeDir);
                 const QString systemFace = QStringLiteral("%1/%2.face.icon").arg(facesDir).arg(user->name);
-                QString accountsServiceFace = QStringLiteral("/var/lib/AccountsService/icons/%1").arg(user->name);
-
-                if (QFile::exists(userFace))
-                    user->icon = QStringLiteral("file://%1").arg(userFace);
+                const QString accountsServiceFace = QStringLiteral("/var/lib/AccountsService/icons/%1").arg(user->name);
+
+                // If the home is encrypted it takes a lot of time to open
+                // up the greeter, therefore we try the system avatar first
+                if (QFile::exists(systemFace))
+                    user->icon = systemFace;
+                else if (QFile::exists(userFace))
+                    user->icon = userFace;
                 else if (QFile::exists(accountsServiceFace))
                     user->icon = accountsServiceFace;
-                else if (QFile::exists(systemFace))
-                    user->icon = QStringLiteral("file://%1").arg(systemFace);
             }
         }
     }
-- 
2.33.0

